quinn/src/connection.rs

```diff
@@ -525,36 +522,42 @@ impl Drop for ConnectionInner {
     }
 }
 
-/// A bidirectional stream, supporting both sending and receiving data.
-///
-/// Similar to a TCP connection. Each direction of data flow can be reset or finished by the
-/// sending endpoint without interfering with activity in the other direction.
+/// A bidirectional stream, consisting of send and receive halves
 pub struct BiStream {
+    /// Send half for this stream
+    pub send: SendStream,
+    /// Receive half for this stream
+    pub recv: RecvStream,
+}
+
+impl BiStream {
+    fn new(conn: ConnectionRef, id: StreamId) -> Self {
+        debug_assert_eq!(id.directionality(), Directionality::Bi);
+        BiStream {
+            send: SendStream::new(conn.clone(), id),
+            recv: RecvStream::new(conn.clone(), id),
+        }
+    }
+}
+
+/// A stream that can only be used to send data
+pub struct SendStream {
     conn: ConnectionRef,
     stream: StreamId,
-
-    // Send only
     finishing: Option<oneshot::Receiver<Option<ConnectionError>>>,
     finished: bool,
-
-    // Recv only
-    // Whether data reception is complete (due to receiving finish or reset or sending stop)
-    recvd: bool,
 }
 
-impl BiStream {
+impl SendStream {
     fn new(conn: ConnectionRef, stream: StreamId) -> Self {
         Self {
             conn,
             stream,
             finishing: None,
             finished: false,
-            recvd: false,
         }
     }
-}
 
-impl Write for BiStream {
     fn poll_write(&mut self, buf: &[u8]) -> Poll<usize, WriteError> {
```

Comment:
This and `poll_finish` should be `pub`, and documented by the comments that were on the removed `trait Write`.

Reply:
Yup.